// @ethernom/ethernom-api/EthernomAPIMgr.js
import {NativeModules, Platform} from 'react-native';
import { DeviceEventEmitter,NativeEventEmitter } from 'react-native';
var Buffer = require('buffer/').Buffer;

module.exports = NativeModules.EtherAPIManager;

ETH_SUCCESS = 0;
ETH_FAIL = -1;

// card errors
ERR_ETH_CARDSCAN_FAILED = 0x0101;
ERR_ETH_CARD_OPEN_DENIED = 0x0102;
ERR_ETH_CARD_OPEN_FAILED = 0x0103;
ERR_ETH_CARD_CONNECTION_LOST = 0x0104;

// service errors
ERR_ETH_SERVICE_OPEN_DENIED = 0x0201;

// js card encoding constants
ETH_STRING_AS_UNICODE =  1;  // DEFAULT FOR STRING
ETH_STRING_AS_HEX 	  =  2;
ETH_STRING_AS_ASUTF8  =  3;
ETH_INT_AS_BYTE       =  4;  // DEFAULT FOR INTS
ETH_INT_AS_2BYTE      =  5;
ETH_INT_AS_4BYTE      =  6;
ETH_AS_ARRAY	      =  7;

ETH_UNICODESTRING  = 0x11;
ETH_UTF8STRING     = 0x12;
ETH_BYTE           = 0x13;
ETH_2BYTE          = 0x14;
ETH_PAYLOAD        = 0x15;
ETH_4BYTE          = 0x16;

// app level constants
ETH_DISCOVERY_COMPLETE = 0x1001;
ETH_SCAN_FOREVER       = 0X1002;

//GENERIC APP COMMAND
SERIVCE_PORT 		   = 0x16;

// mobile to card
CM_LAUNCH_APP     = 0x81;
CM_SUSPEND_APP    = 0x82;

// card to mobile
CM_RSP            = 0x01;

// error code
CM_ERR_SUCCESS          = 0x00;
CM_ERR_CARD_BUSY        = 0x01;
CM_ERR_INVALID_CMD      = 0x02;
CM_ERR_APP_NOT_ALLOWED  = 0x04;
CM_ERR_INVALID_IMG_ID   = 0x08;
CM_ERR_APP_BUSY         = 0x09;
CM_ERR_APP_DISCONNECT   = 0x0E;

var instanceNumber = 0, g_iosEventEmitter = null;
if(Platform.OS == 'ios'){
	g_iosEventEmitter = new NativeEventEmitter(NativeModules.EtherAPIManager);
}

module.exports = class EthernomAPI {
	CM_ERR_RESTORE_INCOMPLETE = 0x12;
	
    manager = NativeModules.EtherAPIManager;
    _appid = 0;
    
    currID   = "";
    currName = "";
    currSN   = ""; 

    constructor(adapterConnectionString, App_ID, secsTimeout, allowDuplicates, stopScanOnWrite, callback){
        this._appid = App_ID;
        this.instNo = instanceNumber;
        if(Platform.OS == 'ios'){
            this.iosEventEmitter = g_iosEventEmitter;
            this.discoverSubscription = null;
            this.disconnectedSubscription = null;
            this.unsolicitedSubscription = null;
        }
        instanceNumber = instanceNumber + 1;
		
		if(typeof App_ID == "number"){
			this.manager._InitAPI(adapterConnectionString, App_ID, secsTimeout, allowDuplicates, stopScanOnWrite, (resultCode, instanceHandle) => {
				if (resultCode == ETH_SUCCESS){
					this.instanceHandle = instanceHandle;
					
					this.unsolicitedEventName = 'onUnsolicitedEvent'+this.instanceHandle;
					this.disconnectedEventName = 'onDisconnected'+this.instanceHandle;
					this.eventDiscoverName = 'onDiscoveredDevice'+this.instanceHandle;
				}
				callback(resultCode);
			});
			
		}else if(typeof App_ID == "string"){
			this.manager._InitDFUAPI(adapterConnectionString, App_ID, secsTimeout, allowDuplicates, stopScanOnWrite, (resultCode, instanceHandle) => {
				if (resultCode == ETH_SUCCESS){
					this.instanceHandle = instanceHandle;
					
					this.unsolicitedEventName = 'onUnsolicitedEvent'+this.instanceHandle;
					this.disconnectedEventName = 'onDisconnected'+this.instanceHandle;
					this.eventDiscoverName = 'onDiscoveredDevice'+this.instanceHandle;
				}
				callback(resultCode);
			});
		}
    }
	
	DisconnectListeners = () => {
		if(Platform.OS == 'ios'){
			if (this.discoverSubscription)
				this.discoverSubscription.remove();
			
			if (this.disconnectedSubscription)
				this.disconnectedSubscription.remove();
			
			if (this.unsolicitedSubscription)
				this.unsolicitedSubscription.remove();
			
			this.iosEventEmitter = null;
		}
	}

	DiscoverDevices = (callback)=>{
		if(this._discoverCallback != null) this._discoverCallback = null;
		this._discoverCallback = callback;
		
		if(Platform.OS == 'ios'){
			if (!this.discoverSubscription)
				this.discoverSubscription = this.iosEventEmitter.addListener(
					this.eventDiscoverName,(event) => {
						if(event.dSerialNum != ""){
							var trim = new Buffer(event.dSerialNum, 'hex');
							trim  = trim.reverse();
							trim = trim.slice(0, trim.length-3);
							event.dSerialNum = trim.toString('hex')
						}
						
                        this._discoverCallback(event.resultCode, event.peripheralID, event.deviceName, event.dSerialNum);
					}
				);
		}else{
			DeviceEventEmitter.addListener(this.eventDiscoverName, (event) => {
				if(event.dSerialNum != ""){
					var trim = new Buffer(event.dSerialNum, 'hex');
					trim  = trim.reverse();
					trim = trim.slice(0, trim.length-1);
					event.dSerialNum = trim.toString('hex')
					if(event.dSerialNum.length != 16){
						event.dSerialNum = "00" + event.dSerialNum;
					}
				}
				
				this._discoverCallback(event.resultCode, event.peripheralID, event.deviceName, event.dSerialNum);
			});
		}
		
		this.manager._DiscoverDevices(this.instanceHandle, this.eventDiscoverName);
	}

	OnCardDisconnected = (callback) => {
		if(this._cardDisconnectedCallback != null) this._cardDisconnectedCallback = null;
		this._cardDisconnectedCallback = callback;
		
		if(Platform.OS == 'ios'){
			if(!this.disconnectedSubscription)
				this.disconnectedSubscription = this.iosEventEmitter.addListener(
					this.disconnectedEventName,(event) => {
						this._cardDisconnectedCallback(event.resultCode);
					}
				);
		}else{
			if(this.disconnectedListener != null) this.disconnectedListener.remove();
			this.disconnectedListener = DeviceEventEmitter.addListener(this.disconnectedEventName, (event) => {
				this._cardDisconnectedCallback(event.resultCode);
			});
		}
		
		this.manager._onCardDisconnected(this.instanceHandle, this.disconnectedEventName);
	}
	
	SubscribeToUnsolictedEvents = (deviceID, deviceName, msgDef, callback) => {
		if(this._unsolicitedCallback != null) this._unsolicitedCallback = null;
		this._unsolicitedCallback = callback;
		
		var msg = JSON.stringify(msgDef);
		if(Platform.OS == 'ios'){
			if (!this.unsolicitedSubscription)
			this.unsolicitedSubscription = this.iosEventEmitter.addListener(
				this.unsolicitedEventName,(event) => {
					this._unsolicitedCallback(deviceID, deviceName, event.resultCode,event.msgJSON);
				}
			);
		}else{
			DeviceEventEmitter.addListener(this.unsolicitedEventName, (event) => {
				this._unsolicitedCallback(deviceID, deviceName, event.resultCode,event.msgJSON);
			});
		}

		this.manager._SubscribeToUnsolictedEvents(this.instanceHandle, msg,this.unsolicitedEventName);
	}

	UnSubscribeToUnsolictedEvents = () => {
		this._unsolicitedCallback = null;
		DeviceEventEmitter.removeListener(this.unsolicitedEventName);
		this.manager._UnSubscribeToUnsolictedEvents(this.instanceHandle, this.unsolicitedEventName);
	}

	Select = (deviceID, deviceSN, deviceName, callback) => {
		this.currID = deviceID;
		this.currSN = deviceSN;
		this.currName = deviceName;
		console.log("SELECT");
		console.log(deviceID);
		console.log(deviceSN);
		console.log(deviceName);
		this.manager._Select(this.instanceHandle, deviceID, deviceName,  callback);
	}

	StopDiscovery = () => {
		this.manager._StopDiscovery(this.instanceHandle);
	}

	CardClose = (callback) => {
		if (!callback){
			callback = (resultCode) => {
			
			};
		}
		this.manager._CardClose(this.instanceHandle, callback);
	}
	
	WriteJSON_Generic = (inObj, outDef, callback) => {
		var s = JSON.stringify(inObj);
		var soutdef = JSON.stringify(outDef);

		this.manager._WriteJSON_Generic(this.instanceHandle, s, soutdef, callback);
    }

    WriteJSON = (inObj, outDef, callback) => {
		var s = JSON.stringify(inObj);
		var soutdef = JSON.stringify(outDef);

		this.manager._WriteJSON(this.instanceHandle, s, soutdef, callback);
    }

    WriteJSON_DL = (inObj, outDef, callback) => {
		var s = JSON.stringify(inObj);
		var soutdef = JSON.stringify(outDef);

		this.manager._WriteJSON_DL(this.instanceHandle, s, soutdef, callback);
    }
	
	LaunchApp = (APP_ID, callback) => {
		var inObj = [{payload: ETH_PAYLOAD}, {command: ETH_BYTE}, {p1: ETH_BYTE}, {len: ETH_2BYTE}, {code: ETH_BYTE}];
		var outObj = [{command: CM_LAUNCH_APP}, {p1: 0}, {len: {encoding:ETH_INT_AS_2BYTE, value: 1}}, {app: APP_ID}]
		this.WriteJSON_Generic(outObj, inObj, (resultCode, msg) =>{
			if(resultCode == ETH_SUCCESS){
				var msg_obj = JSON.parse(msg);
				if(msg_obj.command == CM_RSP){
					switch(msg_obj.code){
						case CM_ERR_SUCCESS:
							callback(ETH_SUCCESS);
							break;
							
						case CM_ERR_APP_BUSY:
							this._HandleAppBusy(APP_ID, callback);
							break;
						
						case this.CM_ERR_RESTORE_INCOMPLETE:
							callback(this.CM_ERR_RESTORE_INCOMPLETE);
							break;
						
						default:
							callback(ETH_FAIL)
							break;
					}
				}else{
					callback(ETH_FAIL);
				}
			}else{
				callback(ETH_FAIL);
			}
		});
	}

	DoAppKeyExchange = (pin, callback) => {
		this.manager._DoAppKeyExchange(this.instanceHandle, pin, (resultCode, msg) =>{
			if(resultCode == ETH_SUCCESS){
				callback(ETH_SUCCESS);
			}else{
				callback(ETH_FAIL);
			}
		});
	}
	
	WriteJSON_Encrypted = (inObj, outDef, doEncrypt, callback) => {
		var s = JSON.stringify(inObj);
		var soutdef = JSON.stringify(outDef);
		
		this.manager._WriteJSON_Encrypted(this.instanceHandle, doEncrypt, s, soutdef, callback);
    }

	
	_HandleAppBusy = (APP_ID, LaunchAppCallback) => {
		this.DisconnectApp(callback = (resultCode) => {
			if(resultCode == ETH_SUCCESS){
				this.LaunchApp(APP_ID, callback = (resultCode) => {
					if(resultCode == ETH_SUCCESS){
						LaunchAppCallback(ETH_SUCCESS);
					}else{
						LaunchAppCallback(ETH_FAIL);
					}
				})
				
			}else if(resultCode == 1){
				this.CardClose(callback = (resultCode) => {});
				this.Select(this.currID, this.currSN, this.currName, callback = (resultCode) => {
					if(resultCode == ETH_SUCCESS){
						LaunchAppCallback(1);
					}else{
						LaunchAppCallback(ETH_FAIL);
					}
				});
			
			}else{
				LaunchAppCallback(ETH_FAIL);
			}
		})
	}
	
	DisconnectApp = (callback) => {
		var inObj = [{command: ETH_BYTE}, {p1: ETH_BYTE}, {len: ETH_2BYTE}, {code: ETH_BYTE}];
		var outObj = [{command: CM_SUSPEND_APP}, {p1: 0}, {len: {encoding:ETH_INT_AS_2BYTE, value: 1}}, {app: 0}]
		this.WriteJSON_Generic(outObj, inObj, (resultCode, msg) =>{
			if(resultCode == ETH_SUCCESS){
				var msg_obj = JSON.parse(msg);
				switch(msg_obj.command){
					case CM_RSP:
						if(msg_obj.code == CM_ERR_SUCCESS){
							callback(ETH_SUCCESS);
						}else if(msg_obj.code == CM_ERR_APP_DISCONNECT){
							callback(1);
						}else{
							callback(ETH_FAIL)
						}
						break;
					
					default:
						callback(ETH_FAIL)
						break;
				}

			}else{
				callback(ETH_FAIL);
			}
		})
	}
}
